### PR TITLE
JIT: Added four SVE `CreateBreak*` APIs

### PR DESCRIPTION
**Pull request by TIHan**
Contributes to https://github.com/dotnet/runtime/issues/99957

Adds:
- CreateBreakAfterMask
- CreateBreakAfterPropagateMask
- CreateBreakBeforeMask
- CreateBreakBeforePropagateMask

### CODE REVIEW

**Comment by TIHan** [comment]
@dotnet/arm64-contrib @kunalspathak this is ready.

stress test results:
> ===================Running default===================
------------------- {} -------------------
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterMask_byte() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterMask_ushort() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterMask_uint() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterMask_ulong() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterMask_sbyte() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterMask_short() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterMask_int() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterMask_long() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterPropagateMask_byte() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterPropagateMask_ushort() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterPropagateMask_uint() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterPropagateMask_ulong() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterPropagateMask_sbyte() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterPropagateMask_short() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterPropagateMask_int() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakAfterPropagateMask_long() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforeMask_byte() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforeMask_ushort() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforeMask_uint() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforeMask_ulong() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforeMask_sbyte() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforeMask_short() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforeMask_int() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforeMask_long() : 19
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforePropagateMask_byte() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforePropagateMask_ushort() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforePropagateMask_uint() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforePropagateMask_ulong() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforePropagateMask_sbyte() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforePropagateMask_short() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforePropagateMask_int() : 22
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_CreateBreakBeforePropagateMask_long() : 22
===================Running jitstress===================
------------------- {'JitMinOpts': '1'} -------------------
------------------- {'JitStress': '1'} -------------------
------------------- {'JitStress': '2'} -------------------
------------------- {'JitStress': '1', 'TieredCompilation': '1'} -------------------
------------------- {'JitStress': '2', 'TieredCompilation': '1'} -------------------
------------------- {'TailcallStress': '1'} -------------------
------------------- {'ReadyToRun': '0'} -------------------
===================Running jitstressregs===================
------------------- {'JitStressRegs': '1'} -------------------
------------------- {'JitStressRegs': '2'} -------------------
------------------- {'JitStressRegs': '3'} -------------------
------------------- {'JitStressRegs': '4'} -------------------
------------------- {'JitStressRegs': '8'} -------------------
------------------- {'JitStressRegs': '0x10'} -------------------
------------------- {'JitStressRegs': '0x80'} -------------------
------------------- {'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStressRegs': '0x2000'} -------------------
===================Running jitstress2-jitstressregs===================
------------------- {'JitStress': '2', 'JitStressRegs': '1'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '2'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '3'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '4'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '8'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x10'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x80'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x2000'} -------------------

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp [hunk 1]
```diff
@@ -2211,6 +2211,20 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                 break;
             }
 
+            case NI_Sve_CreateBreakAfterMask: 
+            case NI_Sve_CreateBreakBeforeMask:
+            {
+                GetEmitter()->emitInsSve_R_R_R(ins, emitSize, targetReg, op1Reg, op2Reg, INS_OPTS_SCALABLE_B);
```
Comment (kunalspathak) [inline src/coreclr/jit/hwintrinsiccodegenarm64.cpp:2217]: For [BRKB](https://docsmirror.github.io/A64/2023-06/brkb_p_p_p.html) and [BRKA](https://docsmirror.github.io/A64/2023-06/brka_p_p_p.html), if we do not pass `insScalableOpts`, it is treated as `BRK* Pd.B, Pg/Z, Pn.B`. How do we generate the merge variant? If we want to support one of those, should merge be the default one? Accordingly we should also update the summary docs and remove the instruction that we are not supporting. We also need to think about what if all `targetReg`, `op1Reg` and `op2Reg` end up to same register or two of them have same register, does it change the operation of `Pg/M` vs. `Pg/Z`. Lastly, we should add test coverage (if not there currently) where we pass same value to `mask` and `srcMask` to simulate such behavior.

**Review by kunalspathak — CHANGES_REQUESTED** [review]
I think there are some considerations to think about the intent of the `CreateBreakBeforeMask` and `CreateBreakAfterMask` and should they do merge operation or zero operation.

File: src/coreclr/jit/hwintrinsiclistarm64sve.h [hunk 2]
```diff
@@ -37,6 +37,10 @@ HARDWARE_INTRINSIC(Sve,           Count16BitElements,
...
+HARDWARE_INTRINSIC(Sve,           CreateBreakBeforeMask,                                            -1,      2,      true,  {INS_sve_brkb,       INS_sve_brkb,       INS_sve_brkb,       INS_sve_brkb,       INS_sve_brkb,       INS_sve_brkb,       INS_sve_brkb,       INS_sve_brkb,       INS_invalid,        INS_invalid},     HW_Category_SIMD,                  HW_Flag_Scalable|HW_Flag_ExplicitMaskedOperation|HW_Flag_ReturnsPerElementMask|HW_Flag_SpecialCodeGen)
+HARDWARE_INTRINSIC(Sve,           CreateBreakBeforePropagateMask,                                   -1,      3,      true,  {INS_sve_brkpb,      INS_sve_brkpb,      INS_sve_brkpb,      INS_sve_brkpb,      INS_sve_brkpb,      INS_sve_brkpb,      INS_sve_brkpb,      INS_sve_brkpb,      INS_invalid,        INS_invalid},     HW_Category_SIMD,                  HW_Flag_Scalable|HW_Flag_ExplicitMaskedOperation|HW_Flag_ReturnsPerElementMask|HW_Flag_SpecialCodeGen)
```
[the remark this answers is not in the document]
Comment (TIHan) [reply src/coreclr/jit/hwintrinsiclistarm64sve.h:43]: I think that API needs a closer look as its signature doesn't correspond well to its instruction based on its parameter names and parameter count.

File: src/coreclr/jit/hwintrinsic.cpp [hunk 3]
```diff
@@ -1942,6 +1946,22 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
...
+            {
+                GenTree* op3 = retNode->AsHWIntrinsic()->Op(3);
+
+                // HWInstrinsic requires a mask for op3
+                if (!varTypeIsMask(op3))
+                {
+                    retNode->AsHWIntrinsic()->Op(3) =
+                        gtNewSimdCvtVectorToMaskNode(TYP_MASK, op3, simdBaseJitType, simdSize);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (TIHan) [reply src/coreclr/jit/hwintrinsic.cpp:1962]: Thanks, missed that.

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp (hunk 1 above)
Comment (TIHan) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:2217]: When looking at the original generated API, there were two C intrinsic variants that it listed:
M variant https://dougallj.github.io/asil/doc/brka_p_p_p_z_8.html
Z variant https://dougallj.github.io/asil/doc/brka_p_p_p_z_8.html

Based on the API signature of our intrinsic, it matches exactly to the Z variant  - so naturally I think that's most likely the correct one. I figured the M variant that was listed is an artifact of the API generation; as other APIs have before. This also leads me to believe that exposing each variant needs to have their own API.

Comment (a74nh) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:2217]: > I figured the M variant that was listed is an artifact of the API generation

Yes, the API generation will have just blindly put them together.

> This also leads me to believe that exposing each variant needs to have their own API.

We can we replicate the zero version using:
```
CndSel(mask, CreateBreakAfterMask(mask, op1), zero)
```

And then in lowering or somewhere, optimise that to a single BRKA with zeroing.

(Or, if I'm wrong and that doesn't work, then yes, ideally add a extra API)

Doesn't need doing now, but could you raise a ticket please.

Comment (TIHan) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:2217]: > We can we replicate the zero version using:

Do you mean "merge"? The API is the zero version currently

Comment (kunalspathak) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:2217]: > could you raise a ticket please.

+1 on that if we decide to skip a variant of the instruction.

Comment (kunalspathak) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:2217]: I agree with just keeping "Z" version. Just update the summary docs.

https://github.com/dotnet/runtime/pull/104184/files#diff-65452cc3b51792a27a23d428011052b18c0b7de4576424a195e0cf5003dc6a9cR987-R988

**Comment by TIHan** [comment]
@kunalspathak Fixed the comments and merged with main, so it's ready

**Review by kunalspathak — APPROVED** [review]
LGTM. Thanks!

**Comment by kunalspathak** [comment]
/ba-g failures are unrelated. superpmi diff failures are related to timeout.

**Pull request merged**